### PR TITLE
[O2B-1191] Add QC flag type update endpoint

### PR DESCRIPTION
**Pull request by xsalonx**
#### I have a JIRA ticket
- [x] branch and/or PR name(s) include(s) JIRA ID
- [x] issue has "Fix version" assigned
- [x] issue "Status" is set to "In review"
- [x] PR labels are selected

Notable changes for users:
- NA

Notable changes for developers:
- Added QC Flag type update endpoint

Changes made to the database:
- NA

### CODE REVIEW

**Comment by xsalonx** [comment]
To be merged after https://github.com/AliceO2Group/Bookkeeping/pull/1447

File: lib/server/controllers/qcFlagTypes.controller.js [hunk 1]
```diff
@@ -106,8 +106,39 @@ const listQcFlagTypesHandler = async (req, res) => {
...
+ */
+const updateQcFlagTypeHandler = async (req, res) => {
+    const validatedDTO = await dtoValidator(
+        DtoFactory.bodyOnly({
+            name: Joi.string().optional(),
+            method: Joi.string().optional(),
+            color: Joi.string().regex(/#[0-9a-fA-F]{6}/).optional().allow(null),
+            bad: Joi.boolean().optional(),
+            archived: Joi.boolean().optional().allow(null),
+        }).concat(Joi.object({ params: Joi.object({ id: Joi.number().required() }) })),
```
Comment (martinboulais) [inline lib/server/controllers/qcFlagTypes.controller.js:121]: No! The purpose of bodyOnly is to be used exactly when you have only body, if you start to happen params it means you are wrongly using it...

File: lib/server/services/qualityControlFlag/QcFlagTypeService.js [hunk 2]
```diff
@@ -95,6 +95,55 @@ class QcFlagTypeService {
...
+            method,
+            bad,
+            color,
+            lastUpdatedById: user.id,
+            archivedAt: archived === true && !qcFlagType.archvied
+                ? new Date()
+                : archived === false
+                    ? null
+                    : undefined,
+        });
```
Comment (martinboulais) [inline lib/server/services/qualityControlFlag/QcFlagTypeService.js:142]: In a transaction

File: lib/server/services/qualityControlFlag/QcFlagTypeService.js [hunk 3]
```diff
@@ -95,6 +95,55 @@ class QcFlagTypeService {
...
+            throw new ConflictError(`A QC flag with ${failureReason} already exists`);
+        }
+
+        await QcFlagTypeRepository.update(qcFlagType, {
+            name,
+            method,
+            bad,
+            color,
+            lastUpdatedById: user.id,
+            archivedAt: archived === true && !qcFlagType.archvied
```
Comment (martinboulais) [inline lib/server/services/qualityControlFlag/QcFlagTypeService.js:137]: Have you tested this? There is a typo, you should have noticed

File: lib/server/routers/qcFlagTypes.router.js [hunk 4]
```diff
@@ -18,6 +18,19 @@ const { rbacMiddleware } = require('../middleware/rbac.middleware.js');
 exports.qcFlagTypesRouter = {
     path: '/qcFlagTypes',
     children: [
+        {
+            path: ':id',
+            children: [
+                {
+                    method: 'get',
+                    controller: QcFlagTypesController.getQcFlagTypeByIdHandler,
+                },
```
Comment (martinboulais) [inline lib/server/routers/qcFlagTypes.router.js:27]: This endpoint should not be part of this PR

File: test/api/qcFlagTypes.test.js [hunk 5]
```diff
@@ -364,4 +364,91 @@ module.exports = () => {
             expect(titleError.detail).to.equal('"body.bad" is required');
         });
     });
+
+    describe('PUT /api/qcFlagTypes/:id', () => {
```
Comment (martinboulais) [inline test/api/qcFlagTypes.test.js:368]: I think Patch would be more relevant

File: lib/server/services/qualityControlFlag/QcFlagTypeService.js [hunk 6]
```diff
@@ -95,6 +95,55 @@ class QcFlagTypeService {
...
+
+        const user = await getUserOrFail({ userId, externalUserId });
+
+        const qcFlagType = await QcFlagTypeRepository.findOne({ where: { id } });
+        if (!qcFlagType) {
+            throw new NotFoundError(`Quality Control Flag Type with this id (${id}) could not be found`);
+        }
+
+        const existingQcFlagType = await QcFlagTypeRepository.findOne({
+            where: { [Op.not]: { id }, [Op.or]: [name ? { name } : {}, method ? { method } : {}] },
```
Comment (martinboulais) [inline lib/server/services/qualityControlFlag/QcFlagTypeService.js:124]: Here you may want to test only properties that are not equal to the existing QC flag type?

**Review by martinboulais — CHANGES_REQUESTED** [review]

File: lib/server/services/qualityControlFlag/QcFlagTypeService.js (hunk 6 above)
Comment (xsalonx) [reply lib/server/services/qualityControlFlag/QcFlagTypeService.js:124]: no, here I check whether some other flag (not id) has same name or method

File: lib/server/routers/qcFlagTypes.router.js (hunk 4 above)
Comment (xsalonx) [reply lib/server/routers/qcFlagTypes.router.js:27]: this was refactoring, I forgot to remove other occurrence of getById route

File: test/api/qcFlagTypes.test.js (hunk 5 above)
Comment (xsalonx) [reply test/api/qcFlagTypes.test.js:368]: applied

File: lib/server/controllers/qcFlagTypes.controller.js (hunk 1 above)
Comment (xsalonx) [reply lib/server/controllers/qcFlagTypes.controller.js]: refactored

File: lib/server/services/qualityControlFlag/QcFlagTypeService.js (hunk 2 above)
Comment (xsalonx) [reply lib/server/services/qualityControlFlag/QcFlagTypeService.js]: added

File: test/api/qcFlagTypes.test.js (hunk 5 above)
Comment (xsalonx) [reply test/api/qcFlagTypes.test.js]: applied

File: lib/server/services/qualityControlFlag/QcFlagTypeService.js (hunk 3 above)
Comment (xsalonx) [reply lib/server/services/qualityControlFlag/QcFlagTypeService.js]: fixed, test extended

File: lib/server/services/qualityControlFlag/QcFlagTypeService.js (hunk 6 above)
Comment (martinboulais) [reply lib/server/services/qualityControlFlag/QcFlagTypeService.js]: Yes, but if someone want to update the QC flag method, and send the full qc flag, including the current name it should work?

Comment (xsalonx) [reply lib/server/services/qualityControlFlag/QcFlagTypeService.js]: yes, `[Op.not]: { id } ` prevents target flag won't be fetched, I renamed the variable to `conflictingQcFlagType`

Comment (martinboulais) [reply lib/server/services/qualityControlFlag/QcFlagTypeService.js]: Indeed
